### PR TITLE
Fix allocation bugs in `allocate_net_gen`

### PR DESCRIPTION
**Pull request by grgmiller**
This PR closes https://github.com/catalyst-cooperative/pudl/issues/2226 by fixing a bug in the allocate_net_gen process. 
This bug was causing issues as described in https://github.com/singularity-energy/open-grid-emissions/issues/271.

Fix missing fuel code issue
- Adds a new function `add_missing_energy_source_codes_to_gens()` that adds energy_source_codes that appear in the `gf` table but not `gens` to `gens`. 
- In some cases, non-zero fuel consumption and net generation is reported in the EIA-923 generation and fuel table that is associated with an energy_source_code that is not associated with that plant-prime mover in the gens table, which would cause these data to get dropped when these two tables are merged. To fix this, for each plant-pm, this function identifies such esc, and adds them to the `gens_at_freq` table as new energy_source_code columns.
- The sub-function `identify_missing_gf_escs_in_gens()` iterates through each prime mover at each plant to identify when there are fuels reported in the gf table for that plant-pm that are not listed in the gens table for that plant pm. This is currently implemented with for loops, which is slow, but I didn't know of a faster way to do this. Suggestions welcome.

Other bug fixes
- Adds the `MISSING_SENTINEL` value to the `net_generation_mwh_g_tbl` column. For some reason, this column had been commented out, which was leading to NaNs appearing in the data when dividing by this column when the value was zero. I un-commented this line.
- In `allocate_net_gen_by_gen_esc()`, we no longer allow `frac_from_g_tbl` to be greater than 100%. This was previously happening when the mwh reported in the g table were greater than the mwh reported in the gf table. However, numbers greater than 100% was causing the `frac_missing_from_g_tbl` to become negative, which was resulting in nonsensical allocations. We implement the same cap on `frac_from_bf_tbl` in `allocate_fuel_by_gen_esc()`.
- In  `allocate_fuel_by_gen_esc()`, when calculating `frac_cap`, the code had been dividing `capacity_mw` by `capacity_mw_unit_fuel`. However, this was resulting in some nonsensical allocations because fuel is being allocated by PM-fuel, not by unit. We changed this to divide by `capacity_mw_pm_fuel` instead. This is consistent with how `frac_cap` is calculated in the `allocate_net_gen_by_gen_esc()` function

Other cosmetic changes
- Rename `adjust_energy_source_codes()` to `adjust_msw_energy_source_codes()` to more precisely describe what the function does

### CODE REVIEW

**Comment by grgmiller** [comment]
@cmgosnell this is a draft pull request partially because I am still verifying that all of the issues are fixed, and also because I'm working from my laptop this week and won't be able to run all of the pre-commit hooks until I get back to my other computer next week. However, I'd still appreciate any feedback you have in the meantime so that I can make any necessary edits.

**Review by cmgosnell — CHANGES_REQUESTED** [review]
hey @grgmiller ! I'm glad you found this issue. bc of the complex nature of this allocation, I think it would be good to add some unit tests about it. Unit tests for these multi-input tables can be pretty tedious to set up, but I think are exactly what we are missing overall and will help us all make sure any update does what we intend.

I mocked one up for this particular plant you identified below. For me, this passed without your edits. So I poked at it a little and I think the problem is actually the 0 in from `ST` from the generation fuel table.

I didn't dig super deeply into your proposed fix yet but the main comment I have is that I think you can probably avoid the need to loop over all of the plants and their energy source. I don't know if this is time intensive for you when running this with just a few years of data, but I suspect this would get super slow with all of the years. If the main intention is to check for missing ESC's, you can probably do something liiiike: 

```
esc_matches = (
    pd.merge(
        df1[id_cols].drop_duplicates()
        df2[id_cols].drop_duplicates(), on=id_cols, how="outer", indicator=True 
    )
)
esc_matches[esc_matches._merge != "both"]
```
OR you  could set the index and use `symmetric_difference`



### example unit test
```
from pudl.analysis.allocate_net_gen import *
from io import StringIO

gens_at_freq_test = pd.read_csv(
    StringIO(
        """
report_date,plant_id_eia,generator_id,prime_mover_code,unit_id_pudl,capacity_mw,fuel_type_count,operational_status,retirement_date,energy_source_code_1,energy_source_code_2,energy_source_code_3,energy_source_code_4,energy_source_code_5,energy_source_code_6,energy_source_code_7,planned_energy_source_code_1,startup_source_code_1,startup_source_code_2,startup_source_code_3,startup_source_code_4
2019-01-01,8023,1,ST,1,556.0,1,existing,nan,SUB,BIT,null,null,nan,nan,nan,nan,DFO,nan,nan,nan
2019-01-01,8023,2,ST,2,556.0,1,existing,nan,SUB,SUB,BIT,nan,nan,nan,nan,DFO,nan,nan,nan
"""
    ),
).pipe(apply_pudl_dtypes, group="eia")
bf_test = pd.read_csv(
    StringIO(
        """
report_date,plant_id_eia,boiler_id,energy_source_code,prime_mover_code,fuel_consumed_mmbtu
2019-01-01,8023,1,DFO,ST,17853.519999999997
2019-01-01,8023,1,RC,ST,27681065.276
2019-01-01,8023,1,SUB,ST,0.0
2019-01-01,8023,2,DFO,ST,17712.999999999996
2019-01-01,8023,2,RC,ST,29096935.279
2019-01-01,8023,2,SUB,ST,0.0
"""
    ),
).pipe(apply_pudl_dtypes, group="eia")
gen_test = pd.read_csv(
    StringIO(
        """
report_date,plant_id_eia,generator_id,net_generation_mwh
2019-01-01,8023,1,2606737.0
2019-01-01,8023,2,2759826.0
"""
    ),
).pipe(apply_pudl_dtypes, group="eia")

bga_test = pd.read_csv(
    StringIO(
        """
plant_id_eia,boiler_id,generator_id,report_date
8023,1,1,2019-01-01
8023,2,2,2019-01-01
"""
    ),
).pipe(apply_pudl_dtypes, group="eia")


gf_test = pd.read_csv(
    StringIO(
        """
report_date,plant_id_eia,energy_source_code,prime_mover_code,net_generation_mwh,fuel_consumed_mmbtu,fuel_consumed_for_electricity_mmbtu
2019-01-01,8023,DFO,ST,3369.286,35566.0,35566.0
2019-01-01,8023,RC,ST,5363193.71,56777578.0,56777578.0
2019-01-01,8023,SUB,ST,0.0, 0.0,0.0
"""
    ),
).pipe(apply_pudl_dtypes, group="eia")

def test_associate_generator_tables():
    gen_assoc_test = associate_generator_tables(
        gens=gens_at_freq_test, gf=gf_test, gen=gen_test, bf=bf_test, bga=bga_test
    )

    assert bf_test.fuel_consumed_mmbtu.sum() == gen_assoc_test.fuel_consumed_mmbtu_bf_tbl.sum()
```

**Comment by jdangerx** [comment]
Hi @grgmiller! @cmgosnell is gonna be out for a bit and I'll try to keep this moving in her stead.

**Comment by grgmiller** [comment]
Thanks for this suggestion @cmgosnell. Your proposed approach to avoid looping through each plant-pm is much more efficient and I have implemented it in my most recent commit. 

Unfortunately, I have no experience writing unit tests so I might need some help from @jdangerx or others on this.

The mock unit test you shared looks like it focuses on the relationship between `bf` and `gens`, while this PR is focused on mismatches between `gf` (not `bf`) and `gens`.

**Comment by jdangerx** [comment]
I edited the test file in `allocate_net_gen.py` with the diff in [this gist](https://gist.github.com/jdangerx/fff60c07b220a8a122e867d3484a2ddf). It passes on `oge_dev`!

There's a bunch of setup code to make a fake `PudlTabl` object which returns the faked dataframes, but the meat of the test is at the bottom - we pass in this fake PudlTabl object to `allocate_gen_fuel_by_generator_energy_source` and then assert that the allocated `fuel_consumed_mmbtu` sums to the same as the pre-allocation `generation_fuel_eia923.fuel_consumed_mmbtu`.

This test also does fail on `dev`, which is a good sign - it actually fails with `100.0% of allocated plant/year's are off by
 more than 5%`, which is cool! We fail if there is some threshold % of plants that fail to allocate accurately - maybe with this fix we can lower that threshold :)

If we stub out that check, the test does end up failing at the assert on `dev`, where we only see the DFO values:
```
>       assert (
            generation_fuel_eia923.fuel_consumed_mmbtu.sum()
            == gen_assoc_test.fuel_consumed_mmbtu.sum()
        )
E       assert 56813144.0 == 35566.0
```

disclaimer / question for people who know this data better than I do: I added a faked `gen_original_eia923` value to `PudlTablMock` which was exactly the same as the non-original one. I think that's OK here, because in this case we already have the generator-level data in the `gen_eia923` table and I'm not expecting any allocation to happen.

File: src/pudl/analysis/allocate_net_gen.py [hunk 1]
```diff
@@ -1598,6 +1605,85 @@ def adjust_energy_source_codes(
...
+    EIA-923 generation and fuel table that is associated with an energy_source_code that
+    is not associated with that plant-prime mover in the gens table, which would cause
+    these data to get dropped when these two tables are merged. To fix this, for each
+    plant-pm, this function identifies such esc, and adds them to the `gens_at_freq`
+    table as new energy_source_code columns.
+    """
+
+    missing_gf_escs_from_gens = identify_missing_gf_escs_in_gens(gens_at_freq, gf)
+
+    # pivot these data to become new numbered energy_source_code_n columns starting at 7
```
Comment (jdangerx) [inline src/pudl/analysis/allocate_net_gen.py:1621]: It starts at 7 because there are 6 defined in the gens table, right? If so, it could be nice to explain that here.

**Review by jdangerx — CHANGES_REQUESTED** [review]
(See the comment about testing I made earlier - just clicking the review button to indicate that I have reviewed this)

**Comment by jdangerx** [comment]
Hi @grgmiller! @cmgosnell and I dug through this a bit earlier today.

We like your approach of adding the new `energy_source_code` to `gens` more than our existing approach where we put mystery data in `*_unasssociated` columns!

We'd love for this pattern to replace the existing pattern - but we need to write tests for the desired behavior, and we should probably handle both "missing energy source codes" as well as "missing primary movers." We're working on a nicer test pattern in #2297 - once that's sorted, we can work together to get your approach merged in.

Finally I'd love to see how much this affects our "off by 5 percent" ratio - just out of curiosity. Maybe we'll be able to bump that acceptance threshold down.

File: src/pudl/analysis/allocate_net_gen.py (hunk 1 above)
Comment (grgmiller) [reply src/pudl/analysis/allocate_net_gen.py:1621]: That's correct. The original energy_source_code are numbered 1-6

**Comment by grgmiller** [comment]
hi @jdangerx sorry for the radio silence here, I've had a lot of other things on my plate recently and am just catching up on this thread. 
We ended up implementing a validation check within the OGE codebase to check that the output data matches the input `gf` data within some threshold. See: https://github.com/singularity-energy/open-grid-emissions/blob/1477ac5cff002cd218bfcc768655b121a83426ac/src/validation.py#L43

I had defined this as an absolute threshold of 10 MWh or 10 mmbtu, but I think it makes more sense to do a percentage. I also think that it makes more sense to implement this validation check within the allocate_net_gen module instead of OGE, but this was just a stopgap so that we could publish our patch release. But this is catching relatively large (and yet unexplained) differences in the allocated total relative to the input total. For example, for 2021, here is the output of our warning (the value shows how much the allocated value is higher (+) or lower (-) than the input gf total for that plant):

```
WARNING: Allocated EIA-923 doesn't match input data for plants:
              net_generation_mwh  fuel_consumed_mmbtu  fuel_consumed_for_electricity_mmbtu
plant_id_eia
1206                   60.568183          1241.475753                          1241.475753
1316                  -48.000000             0.000000                             0.000000
52148                   1.601000            18.000000                             7.000000
54224                  14.960000           150.000000                            66.000000
54967                   1.960000            24.000000                            24.000000
58256                -241.000000             0.000000                             0.000000
59260                 836.856000         13729.000000                         13729.000000
59825                9539.000000         84354.000000                         84354.000000
```

This likely indicates that there are still some allocation bugs that we need to fix somewhere, since these differences are definitely higher than rounding errors. 

Let me know what I can do to help move this along. I know you had asked about writing unit tests last week. Not sure if you already did that or that still needs to happen.

**Comment by cmgosnell** [comment]
@grgmiller  thanks for this extra context! totally agree that there are still hidden bugs in here. When i was trying to get this all integrated last year, there were definitely outliers and issues. I think added more tests cases will help make the process easier to debug. My experience with it was: find an unexpected behavior, try to fix that, end up making  the overall results worse or introduce more or different unexpected behaviors. For me at least, a looooot of the weirdness happened because of 0's. I am still pretty tempted to null out and/or drop all zeros. But again, adding tests for these edge cases to ensure we get the expected outcomes and also don't mess up other cases in the process seems like the key here.

@jdangerx  added some new tests and cleaned up the test in #2297. that's all ready to go imo so hopefully we can merge it into `dev`.

As @jdangerx said, I really like this method of adding the ESC's into the gens table before the allocation. I think that this should remove the need for the `_allocate_unassociated_records` for the ESC. which is great because the un-associated records also introduced a ton of unexpected outcomes. I believe with your new `add_missing_energy_source_codes_to_gens` we should be able to remove dealing with the un-associated ESC records [here](https://github.com/catalyst-cooperative/pudl/blob/167ee893288e7dbe814bb08347eb007ae0521c08/src/pudl/analysis/allocate_net_gen.py#L638). This would remove some of the columns in the dfs which would also be nice.

In order to move this forward, I'd propose integrating the changes you've made so far (+ maybe taking out the unassociated ESC record step if your up for it) and then making a test/bug squashing plan so we can break out managible tasks here.

**Comment by grgmiller** [comment]
Thanks @cmgosnell - just to clarify is there anything that you need me to do to help advance this on my end, or is @jdangerx taking the lead on this?

**Comment by cmgosnell** [comment]
I'm going to suggest that we merge this in with the following scope and move the remaining tasks into #1113:

``` [tasklist]
### In scope:
- [ ] Remove the now duplicated `_allocate_unassociated_records` for the energy source codes
```

### Out-of-scope
- [ ] Add more tests & debug errors (will flesh this out in #1113

**Comment by jdangerx** [comment]
Closing in favor of #2446
